File: code/brf_spr.r
# black rockfish selectivity at age
# ben.williams@alaska.gov

# load ----
source("code/helper.r")

setwd(here::here("code"))

# data ----

# note if the lengths are not rounded there are occasionaly more precise estimates
# these will not allow the model to converge
# combine sport and comm age data

# Female only data for sport and commercial fisheries
read_csv(here::here("data/br_bio.csv"), guess = 50000) %>% 
  rename_all(tolower) %>% 
  dplyr::select(year, Area = g_management_area_code, 
                length = length_millimeters, age, sex = sex_code,
                weight = weight_kilograms) %>% 
  filter(sex == 2) %>%
  dplyr::select(-sex) %>%
  mutate(length = round(length / 10),
         fish = "comm") %>% 
  bind_rows(read_csv(here::here("data/sport_brf_bio_se.csv"), guess = 50000) %>% 
              rename_all(tolower) %>% 
              filter(sex=="F") %>%
              dplyr::select(year, Area = area, 
                            length, age, sex , weight = wt_kg) %>% 
              dplyr::select(-sex) %>%
              mutate(length = round(length / 10),
                     fish = "sport")) %>% 
  filter(Area %in% c("CSEO"), !is.na(age), !is.na(length), 
         year!=2001) -> brf

# read_csv(here::here("data/kodiak_brf.csv"), guess = 50000) %>%
#   filter(age>0 & age<26, sex==2, !is.na(age), !is.na(length),
#          year>2001) -> brf

# estimate length/weight relationship

lw <- unname(lm(log(weight) ~ log(length), data = brf)$coef)
lw_int = (lw[1])
lw_slope = (lw[2])

data.frame(length = 1:70) %>% 
  mutate(weight = exp(lw_int + lw_slope * log(length)))  %>% 
  ggplot(aes(length, weight)) + 
  geom_line() + 
  geom_point(data = brf, alpha = 0.2)


# maturity at age from Kodiak study

mat_int = -7.521637
mat_slope = 0.717806

data.frame(age = 1:25) %>% 
  mutate(maturity = exp(mat_int + mat_slope * age) / (1 + exp(mat_int + mat_slope * age)))  %>% 
  ggplot(aes(age, maturity)) + 
  geom_line()

# choose data ----
# length based
# 
# dat <- clean_up(brf, length) %>%
#   left_join(data.frame(X = 20:65), .) %>%
#   mutate(prop = replace_na(prop, 0)) -> dat

# age based 

clean_up(brf, age) %>% 
  # group_by(year) %>% 
  left_join(expand.grid(X = 1:30), .) %>%
  mutate(prop = replace_na(prop, 0)) -> dat

# pre-flight check that the data are in a good form and starting values are reasonable 
# von b
Linf <- 54
kappa <- 0.15
t0 <- -0.5

brf %>% 
  mutate(fit = Linf * (1 - exp(-kappa * (age - t0)))) %>% 
  ggplot(aes(age, length)) + 
  geom_point() + 
  geom_line(aes(y = fit)) +
  expand_limits(y = 0)

# selectivity

alpha = 10
beta = -1.0
M = 0.123
F = 0.07

dat %>% 
  mutate(sel =  1 / (1 + exp(alpha + beta * X ))) -> dd

dd$N <- 1

for(i in 2:nrow(dd)){
  dd$N[i] = dd$N[i-1] * exp(-M -F * dd$sel[i-1])
}

dd %>%
  # group_by(year) %>% 
  mutate(C = N * (1 - exp(-M -F * sel)) * F * sel/(M + F * sel),
         fit_prop = C / max(C)) %>% 
  ggplot(aes(X, prop)) + 
  geom_bar(stat = "identity", alpha = 0.3) +
  geom_line(aes(y = fit_prop)) 

# TMB model ----
compile("brf_spr.cpp")
dyn.load(dynlib("brf_spr"))

# data 
data = list(age = brf$age,
            length = brf$length,
            X = dat$X, 
            prop = dat$prop,
            lw_int = lw_int,
            lw_slope = lw_slope,
            mat_b0 = mat_int,
            mat_b1 = mat_slope,
            mu_M = 0.13, # based on barefoot ecologist tool
            sigma_M = 0.01)

# starting parameters
params <- list(Linf = 55, kappa = 0.15, t0 = -0.5, log_von_sigma = 0.001,
               Fcur = 0.07, M = 0.13,
               alpha = 10, beta = -1.0, log_sel_sigma = 0.0001)

# add parameter bounds 
# map = list(M = factor(NA))
           # alpha = factor(NA), beta = factor(NA)) #M = factor(NA)
map = list()

L = list(Linf = 20, kappa = 0.05, t0 = -10.0, log_von_sigma = 0.0001,
         Fcur = 0.03, 
         M = 0.05,
         alpha = 5, 
         beta = -2.5, 
         log_sel_sigma = 0.0001)

U = list(Linf = 70, kappa = 0.35, t0 = 5.0, log_von_sigma = 10,
         Fcur = 0.2, 
         M = 0.3,
         alpha = 17, 
         beta = -0.9, 
         log_sel_sigma = 10)

# build model
model <- MakeADFun(data = data, 
                   parameters = params, 
                   DLL="brf_spr",
                   map = map)

# optimize the model
fit <- nlminb(model$par, 
              model$fn, 
              model$gr,
              upper = U,
              lower = L)

best <- model$env$last.par.best
rep <- sdreport(model)
rep

best
summary(rep, select = "all", p.value = T)


yfit <- model$report()$yfit
fit_prop <- model$report()$fit_prop
a <- model$report()$alpha
b <- model$report()$beta
N <- model$report()$Na
unN <- model$report()$UnNa
S <- model$report()$Sa
C <- model$report()$Ca
M <- model$report()$M
Fcur <- model$report()$Fcur
fished <- model$report()$fished 
unfished <- model$report()$unfished 
model$report()$exploitB

ifelse(is.nan(fished), 0, fished) -> fished
ifelse(is.nan(unfished), 0, unfished) -> unfished

data.frame(age = dat$X,
           fished = fished, 
           unfished = unfished) %>% 
  ggplot(aes(age, unfished)) + 
  geom_line() + 
  geom_line(aes(y = fished), color = 4) + 
  expand_limits(y = c(0, 0.03))
sum(fished) / sum(unfished) # SPR

# von b

brf %>% 
  mutate(fit = yfit) %>% 
  ggplot(aes(age, length)) + 
  geom_point(alpha = 0.1) + 
  geom_line(aes(y = yfit)) + 
  expand_limits(x = 0, y = 0)

# age comp 
dat %>% 
  mutate(sel = S,
         catch = C,
         numbers = N,
         fit = fit_prop) %>% 
  ggplot(aes(X, prop)) + 
  geom_bar(stat = "identity", alpha = 0.3) +
  geom_line(aes(y = fit), color = 1)

fem_A50 <- -a / b # 50% selected
# female <- rep
# fem_M <- M
# male_A50 <- -a / b
# male <- rep
# male_M <- M

File: code/brf_spr.cpp
// determine selectivity parameters conditioned on catch at age
// determine current SPR level

#include<TMB.hpp>
template<class Type>

Type objective_function<Type>::operator()()
{
  // data -----
  
  // von b
  DATA_VECTOR(age);
  DATA_VECTOR(length);
  int von_n = age.size();
  
  //selectivity
  DATA_VECTOR(X);
  DATA_VECTOR(prop);
  int n = X.size();
  
  // weight & maturity

  DATA_INTEGER(lw_int);
  DATA_INTEGER(lw_slope);
  DATA_INTEGER(mat_b0);
  DATA_INTEGER(mat_b1);
  
  // natural mortality prior (assume normal prior with mu and sigma)
  DATA_SCALAR(mu_M)           
  DATA_SCALAR(sigma_M)        

  // parameters ----
  
  //von b
  PARAMETER(Linf);
  PARAMETER(kappa);
  PARAMETER(t0)
  PARAMETER(log_von_sigma);
  
  PARAMETER(Fcur);
  PARAMETER(M);
  PARAMETER(alpha);
  PARAMETER(beta);
  PARAMETER(log_sel_sigma);
  
  // procedures (transform parameters) -----
  Type von_sigma = exp(log_von_sigma);
  Type sel_sigma = exp(log_sel_sigma);
  
  // age-based natural mortality M
  vector<Type> L(n);
  vector<Type> log_L(n);
  //vector<Type> M(n);
  //Type log_Linf = log(Linf);
  //Type log_kappa = log(kappa);

  // vectors for predictions
  // von b
  vector<Type> yfit(von_n); 
  
  // selectivity
  vector<Type> Na(n);
  vector<Type> UnNa(n); //unfished abundance
  vector<Type> Sa(n);
  vector<Type> Ca(n);
  vector<Type> fit_prop(n);
  
  // fishing mortality at age
  vector<Type> Fa(n);
  vector<Type> expN(n);
  vector<Type> totBio(n);
  vector<Type> expBio(n);

  // weight 
  vector<Type> Wa(n);
  vector<Type> Mat(n);
  vector<Type> CWa(n);
  
  vector<Type> unfished(n);
  vector<Type> fished(n); 
  
  // negative log likelihood - set to 0 ----
  // von b
  Type von_nll = 0.0;
  
  // selectivity
  Type nll = 0.0;
  Type nll1 = 0.0;
 
  // total nll
  Type tot_nll = 0.0;
  
  // von b -------------------------------------------------------------
  for(int i=0; i<von_n; i++){
    yfit(i) = Linf * (Type(1.0) - exp(-kappa * (age(i) - t0)));
  }
  
  von_nll = -sum(dnorm(length, yfit, von_sigma, true));
  
  // M --------------------------------------------------------------------

  for(int i=0; i<n; i++){
    L(i) = Linf * (Type(1.0) - exp(-kappa * (X(i) - t0)));
  }

  for(int i=0; i<n; i++){
  log_L(i) = log(L(i));
	}
  
// for(int i=0; i<n; i++){
//  M(i) = exp(Type(0.55) - Type(1.61) * log_L(i) + Type(1.44) * log_Linf + log_kappa);
//  }
  
  // weight at age
  for(int i=0; i<n; i++){
  	Wa(i) = exp(lw_int + lw_slope * log_L(i));
  }

  
  //selectivity ----------------------------------------------------------------
  
  for(int i=0; i<n; i++){
    Sa(i) = Type(1.0) / (Type(1.0) + exp(alpha + beta * X(i) ));
  } 
  
  nll = -sum(dnorm(prop, Sa, sel_sigma, true));
  
//  for(int i=0; i<n; i++){
//  	Fa(i) = Sa(i) * Fcur;
//  }
 
  for(int j=0; j<n; j++){
    
    if(j == 0) Na(j) = 1000;
    if(j > 0) 
      Na(j) = Na(j-1) * exp(-(M + Sa(j-1) * Fcur));
  }

  for(int j=0; j<n; j++){
    
    if(j == 0) UnNa(j) = 1000;
    if(j > 0) 
      UnNa(j) = UnNa(j-1) * exp(-M);
  }

  // maturity at age -----------------------------------------------------------

  for(int i=0; i<n; i++){
  	Mat(i) = exp(mat_b0 + mat_b1 * X(i)) / (Type(1.0) + exp(mat_b0 + mat_b1 * X(i)));
  }

  // catch at age --------------------------------------------------------------

  for(int i=0; i<n; i++){
    Ca(i) = Na(i) * (1 - exp(-(M + Sa(i) * Fcur))) * Fcur * Sa(i) / (M + Sa(i) * Fcur);
   }
  
  Type maxCa = max(Ca);
  
  for(int i = 0; i<n; i++){
    fit_prop(i) = Ca(i) / maxCa; 
  }
  
  nll1 = -sum(dnorm(prop, fit_prop, sel_sigma, true));

  // Prior for natural mortality
  Type prior_M = 0;
  prior_M += Type(0.5) * pow(M - mu_M, 2) / pow(sigma_M, 2);
  
  // biomass estimates --------------------
	for(int i = 0; i<n; i++){
		CWa(i) = Ca(i) * Wa(i);
		unfished(i) = UnNa(i) * Wa(i) * Mat(i);
		fished(i) = Na(i) * Wa(i) * Mat(i);
		Fa(i) = Sa(i) * Fcur;
		expN(i) = Na(i) * Sa(i);
		totBio(i) = Na(i) * Wa(i);
		expBio(i) = expN(i) * Wa(i);
	}
  
  Type exploitN = sum(Ca) / sum(expN);
	Type exploitB = sum(CWa) / sum(expBio);
  
  tot_nll += nll;
  tot_nll += nll1;
  tot_nll += von_nll;
  tot_nll += prior_M;
  
  // reports ---- --------------------------------------------------------------------
  
  //von b
  
  REPORT(Linf);
  REPORT(kappa);
  REPORT(t0);
  REPORT(yfit);
  REPORT(von_sigma);
  REPORT(von_nll);
  
  REPORT(M);
  //REPORT(Fa);
  REPORT(Mat);
  
  REPORT(alpha);
  REPORT(beta);
  REPORT(sel_sigma);
  REPORT(nll);
  REPORT(nll1);
  REPORT(fit_prop);
  REPORT(Sa);
  REPORT(Na);
  REPORT(UnNa);
  REPORT(Ca);
  REPORT(Fcur);
  REPORT(M);
  REPORT(CWa);
  REPORT(unfished);
  REPORT(fished);
  REPORT(exploitN);
  REPORT(exploitB)
  
  return tot_nll;
}


